Improve Cascade CLI

jedimahdi opened the pull request:
This PR resolves #10, #12, and #14.

jedimahdi committed:
feat(cascade-cli): add http-port and version option to cli

jedimahdi committed:
feat(cascade-cli): add postgres options

the-dr-lazy commented on cascade-cli/src/Cascade/CLI.hs:
```diff
@@ -0,0 +1,92 @@
+cascadeVersion version = intercalate "\n" [cVersion, cHash, cDate]
+ where
+  cVersion, cHash, cDate :: String
+  cVersion = "Cascade CLI " <> "v" <> showVersion version
```
"Cascade CLI " <> "v" = "Cascade CLI v"

the-dr-lazy commented on cascade-cli/src/Cascade/CLI.hs:
```diff
@@ -0,0 +1,92 @@
+cliParser :: Version -> ParserInfo Options
+cliParser version = info (helper <*> versionP version <*> cascadeP) $ fullDesc <> progDesc "Cascade Cli"
+
+cascadeCli :: IO ()
+cascadeCli = execParser (cliParser Meta.version) >>= runCascadeApi
```
This is the main CLI function so maybe we can call it `Cascade.CLI.main`.

the-dr-lazy commented on cascade-cli/src/Cascade/CLI.hs:
```diff
@@ -0,0 +1,92 @@
+
+data Options = Options
+  { httpPort         :: !Int
+  , postgresHost     :: !ByteString
+  , postgresPort     :: !Int
+  , postgresUser     :: !ByteString
+  , postgresPassword :: !ByteString
+  , postgresDatabase :: !ByteString
```
We can group these fields into one record. It'll look more convenient.

the-dr-lazy commented on cascade-cli/src/Cascade/CLI.hs:
```diff
@@ -0,0 +1,92 @@
+cascadeP =
+  Options
+    <$> option auto (long "http-port" <> metavar "INT" <> value 3141 <> showDefault <> help "Port number of Cascade Api")
+    <*> strOption (long "postgres-host" <> metavar "STRING" <> value "localhost" <> showDefault <> help "Postgresql host")
```
Just a format suggestion:

```
:%s/Postgresql/PostgreSQL
```

the-dr-lazy commented on cascade-cli/src/Cascade/CLI.hs:
```diff
@@ -0,0 +1,92 @@
+  , postgresDatabase :: !ByteString
+  }
+
+cascadeP :: Parser Options
```
This is actually `optionsP`.

jedimahdi committed:
refactor(cascade-cli): refactor and add PostgresOption and Config data types

the-dr-lazy commented on cascade-cli/src/Cascade/CLI.hs:
```diff
@@ -0,0 +1,99 @@
+runCascadeApi :: Options -> IO ()
+runCascadeApi Options {..} = do
+  databaseConnectionPool <- mkDatabaseConnectionPool postgresOptions
+  Cascade.Api.main <| Cascade.Api.Config { port = httpPort, withDatabaseConnection = Pool.withResource databaseConnectionPool }
```
pipe operator is redundant.

the-dr-lazy commented on cascade-cli/src/Cascade/CLI.hs:
```diff
@@ -0,0 +1,99 @@
+      <*> strOption
+            (mconcat
+              [long "postgres-database", metavar "CASCADE_POSTGRES_DATABASE", value "cascade-api", showDefault, help "PostgreSQL database"]
+            )
```
This one can move to the top binding and be named `postgresOptionsP`.

Also, the parsers can be defined in a do notation:

```hs
postgresOptionsP :: Parser PostgresOptions
postgresOptionsP = do
  host <- strOption <| mconcat [ long "postgres-host"
                               , metavar "CASCADE_POSTGRES_HOST"
                               , value "localhost"
                               , showDefault
                               , help "PostgreSQL host"
                               ]
  ...
  
  pure PostgresOptions { .. }
```

the-dr-lazy commented on cascade-cli/src/Cascade/CLI.hs:
```diff
@@ -0,0 +1,99 @@
+  postgresOpts =
+    PostgresOptions
+      <$> strOption (mconcat [long "postgres-host", metavar "CASCADE_POSTGRES_HOST", value "localhost", showDefault, help "PostgreSQL host"])
+      <*> option auto (mconcat [long "postgres-port", metavar "CASCADE_POSTGRES_PORT", value 5432, showDefault, help "Postgres port"])
```
```diff
- Postgres port
+ PostgreSQL port
```

the-dr-lazy commented on cascade-cli/src/Cascade/CLI.hs:
```diff
@@ -0,0 +1,99 @@
+versionP version = infoOption (cascadeVersion version) <| mconcat [long "version", short 'v', help "Show cascade's version"]
+
+data PostgresOptions = PostgresOptions
+  { postgresHost     :: !String
+  , postgresPort     :: !Word16
+  , postgresUser     :: !String
+  , postgresPassword :: !String
+  , postgresDatabase :: !String
+  }
```
`postgres` prefix is redundant.

the-dr-lazy commented on cascade-cli/src/Cascade/CLI.hs:
```diff
@@ -0,0 +1,99 @@
+  databaseConnectionPool <- mkDatabaseConnectionPool postgresOptions
+  Cascade.Api.main <| Cascade.Api.Config { port = httpPort, withDatabaseConnection = Pool.withResource databaseConnectionPool }
+
+cascadeVersion :: Version -> String
```
There is no need to get `Version` as an input. We can simply use `Meta.version`.

jedimahdi committed:
refactor(cascade-cli): seprate function for postgres options and fix tests

jedimahdi committed:
style(cascade-cli): run brittany

jedimahdi committed:
chore: add cascade-cli/src to headroom source paths

the-dr-lazy committed:
refactor(cascade-cli): remove strict field notations

As the -XStrictData language extension is enabled there is no need to
notate strictness of product types.

the-dr-lazy committed:
feat(cascade-cli): align version information columns

the-dr-lazy committed:
refactor(cascade-cli): rename cliParser to cliP
